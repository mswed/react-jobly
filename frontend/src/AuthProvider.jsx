import React, { useState, useEffect } from 'react';
import JoblyApi from './api';

// We create the context
const AuthContext = React.createContext();

// We set up a component to manage the context
const AuthProvider = ({ children }) => {
  // We initialize the token with the value from localStorage
  const [token, setToken] = useState(localStorage.getItem('jobly-token'));

  // Whenever the token changes (login/logout) we need to update
  // both the api module and localStorage
  useEffect(() => {
    JoblyApi.token = token;
    localStorage.setItem('jobly-token', token || '');
  }, [token]);

  const login = async (username, password) => {
    try {
      const token = await JoblyApi.login(username, password);
      setToken(token);
      return { success: true };
    } catch (error) {
      return { success: false, error };
    }
  };

  const register = async (username, password, firstName, lastName, email) => {
    try {
      const token = await JoblyApi.register(username, password, firstName, lastName, email);
      setToken(token);
      return { success: true };
    } catch (error) {
      return { success: false, error };
    }
  };

  // Create the value object
  const value = { token, register };

  // Wrap everything in the context
  return <AuthContext.Provider value={value}>{children}</AuthContext.Provider>;
};

export { AuthContext, AuthProvider };
