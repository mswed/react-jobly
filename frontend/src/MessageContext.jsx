import React, { useState } from 'react';

// We create the context
const MessageContext = React.createContext();

// We set up a component to manage the context
const MessagesProvider = ({ children }) => {
  const [message, setMessage] = useState(null);

  const showMessage = (text, variant = 'success', duration = 5000) => {
    setMessage({ text, variant, duration });
    // Auto clear
    if (duration) {
      setTimeout(() => {
        clearMessage();
      }, duration);
    }
  };

  const clearMessage = () => {
    setMessage(null);
  };

  return <MessageContext.Provider value={{ message, showMessage, clearMessage }}>{children}</MessageContext.Provider>;
};

export { MessageContext, MessagesProvider };
