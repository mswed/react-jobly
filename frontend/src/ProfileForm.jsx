import { Form, Button } from 'react-bootstrap';
import { useState, useContext, useEffect } from 'react';
import { AuthContext } from './AuthProvider';
import JoblyApi from './api';
import { validateForm } from './utils/formValidation';

const ProfileForm = () => {
  const INITIAL_STATE = {
    firstName: '',
    lastName: '',
    email: '',
  };

  // Get functions from AuthContext
  const { token, currentUser } = useContext(AuthContext);

  // Manage the form inputs
  const [profileForm, setProfileForm] = useState(INITIAL_STATE);

  const validations = {
    firstName: ['required'],
    lastName: ['required'],
    email: ['required'],
  };
  const handleChange = (evt) => {
    const { name, value } = evt.target;
    setProfileForm((originalData) => ({ ...originalData, [name]: value }));
  };

  const handleSubmit = async (evt) => {
    evt.preventDefault();
    validateForm(profileForm, validations);
    await JoblyApi.updateUser(currentUser, profileForm.firstName, profileForm.lastName, profileForm.email);
  };

  useEffect(() => {
    if (currentUser && token && token.trim !== '') {
      // We only run if we have a token and a user
      async function getUser() {
        try {
          const res = await JoblyApi.getUser(currentUser);
          setProfileForm({
            firstName: res.firstName,
            lastName: res.lastName,
            email: res.email,
          });
        } catch (error) {
          console.error('Error fetching user', error);
        }
      }
      getUser();
    }
  }, [token, currentUser]);

  return (
    <Form onSubmit={handleSubmit}>
      <Form.Group className="mb-3" controlId="username">
        <Form.Label>Username</Form.Label>
        <Form.Control placeholder="Username" name="username" value={currentUser} disabled />
      </Form.Group>
      <Form.Group className="mb-3" controlId="first-name">
        <Form.Label>First Name</Form.Label>
        <Form.Control placeholder="First Name" name="firstName" onChange={handleChange} value={profileForm.firstName} />
      </Form.Group>
      <Form.Group className="mb-3" controlId="last-name">
        <Form.Label>Last Name</Form.Label>
        <Form.Control placeholder="Last Name" name="lastName" onChange={handleChange} value={profileForm.lastName} />
      </Form.Group>
      <Form.Group className="mb-3" controlId="email">
        <Form.Label>Email</Form.Label>
        <Form.Control placeholder="Email" name="email" onChange={handleChange} value={profileForm.email} />
      </Form.Group>
      <Button type="submit" className="w-100">
        Save Changes
      </Button>
    </Form>
  );
};

export default ProfileForm;
