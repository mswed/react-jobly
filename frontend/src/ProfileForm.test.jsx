import { it, expect, vi, describe, beforeEach } from 'vitest';
import { render, screen, fireEvent, waitFor } from '@testing-library/react';
import ProfileForm from './ProfileForm';
import { MemoryRouter } from 'react-router-dom';
import { MessageContext } from './MessageContext';
import { AuthContext } from './AuthProvider';
import JoblyApi from './api';

// Mock authentication
const authValue = {
  token: 'fake-token',
  currentUser: 'testuser',
  logout: vi.fn(),
};

// Mock the api
vi.mock('./api', () => ({
  default: {
    getUser: vi.fn().mockResolvedValue({
      firstName: 'Test',
      lastName: 'User',
      email: 'test@example.com',
    }),
    updateUser: vi.fn().mockResolvedValue({}),
  },
}));

describe('ProfileForm', () => {
  // Reset all mocks before each test
  beforeEach(() => {
    vi.clearAllMocks();
  });

  it('renders without crashing', () => {
    render(
      <MemoryRouter>
        <MessageContext.Provider value={{ showMessage: vi.fn() }}>
          <AuthContext.Provider value={authValue}>
            <ProfileForm />
          </AuthContext.Provider>
        </MessageContext.Provider>
      </MemoryRouter>
    );
  });

  it('matches snapshot', () => {
    const { asFragment } = render(
      <MemoryRouter>
        <MessageContext.Provider value={{ showMessage: vi.fn() }}>
          <AuthContext.Provider value={authValue}>
            <ProfileForm />
          </AuthContext.Provider>
        </MessageContext.Provider>
      </MemoryRouter>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it('has all the expected fields and a submit button', () => {
    render(
      <MemoryRouter>
        <MessageContext.Provider value={{ showMessage: vi.fn() }}>
          <AuthContext.Provider value={authValue}>
            <ProfileForm />
          </AuthContext.Provider>
        </MessageContext.Provider>
      </MemoryRouter>
    );

    // Check for form elements
    expect(screen.getByLabelText(/Username/i)).toBeInTheDocument();
    expect(screen.getByLabelText(/First Name/i)).toBeInTheDocument();
    expect(screen.getByLabelText(/Last Name/i)).toBeInTheDocument();
    expect(screen.getByLabelText(/Email/i)).toBeInTheDocument();
    expect(screen.getByRole('button', { name: /Save Changes/i })).toBeInTheDocument();
  });

  it('updates form state when typing', () => {
    render(
      <MemoryRouter>
        <MessageContext.Provider value={{ showMessage: vi.fn() }}>
          <AuthContext.Provider value={authValue}>
            <ProfileForm />
          </AuthContext.Provider>
        </MessageContext.Provider>
      </MemoryRouter>
    );

    // Get form inputs
    const firstNameInput = screen.getByLabelText(/First Name/i);
    const lastNameInput = screen.getByLabelText(/Last Name/i);
    const emailInput = screen.getByLabelText(/Email/i);

    // Type in inputs
    fireEvent.change(firstNameInput, { target: { value: 'Test' } });
    fireEvent.change(lastNameInput, { target: { value: 'User' } });
    fireEvent.change(emailInput, { target: { value: 'test.user@test.com' } });

    // Check values were updated
    expect(firstNameInput).toHaveValue('Test');
    expect(lastNameInput).toHaveValue('User');
    expect(emailInput).toHaveValue('test.user@test.com');
  });

  it('calls updateUser with form data on submit', async () => {
    // Mock functions
    const mockSubmit = vi.fn().mockResolvedValue({ success: true });
    const mockShowMessage = vi.fn();

    render(
      <MemoryRouter>
        <MessageContext.Provider value={{ showMessage: mockShowMessage }}>
          <AuthContext.Provider value={authValue}>
            <ProfileForm />
          </AuthContext.Provider>
        </MessageContext.Provider>
      </MemoryRouter>
    );

    // Get form inputs
    const firstNameInput = screen.getByLabelText(/First Name/i);
    const lastNameInput = screen.getByLabelText(/Last Name/i);
    const emailInput = screen.getByLabelText(/Email/i);

    // Fill form
    fireEvent.change(firstNameInput, { target: { value: 'Test' } });
    fireEvent.change(lastNameInput, { target: { value: 'User' } });
    fireEvent.change(emailInput, { target: { value: 'test.user@test.com' } });

    // Submit form
    fireEvent.click(screen.getByRole('button', { name: /Save Changes/i }));

    // Check login was called with correct data
    await waitFor(() => {
      expect(JoblyApi.updateUser).toHaveBeenCalledWith('testuser', 'Test', 'User', 'test.user@test.com');
    });
  });

  it('shows success message on successful update', async () => {
    // Mock functions for successful login
    const mockSubmit = vi.fn().mockResolvedValue({ success: true });
    const mockShowMessage = vi.fn();

    render(
      <MemoryRouter>
        <MessageContext.Provider value={{ showMessage: mockShowMessage }}>
          <AuthContext.Provider value={authValue}>
            <ProfileForm />
          </AuthContext.Provider>
        </MessageContext.Provider>
      </MemoryRouter>
    );

    // Get form inputs
    const firstNameInput = screen.getByLabelText(/First Name/i);
    const lastNameInput = screen.getByLabelText(/Last Name/i);
    const emailInput = screen.getByLabelText(/Email/i);

    // Fill form
    fireEvent.change(firstNameInput, { target: { value: 'Test' } });
    fireEvent.change(lastNameInput, { target: { value: 'User' } });
    fireEvent.change(emailInput, { target: { value: 'test.user@test.com' } });

    // Submit form
    fireEvent.click(screen.getByRole('button', { name: /Save Changes/i }));

    // Check results
    await waitFor(() => {
      expect(mockShowMessage).toHaveBeenCalledWith('Updated user profile!', 'success');
    });
  });

  it('shows error message on failed update', async () => {
    // Mock functions for failed login
    const mockSubmit = vi.fn().mockResolvedValue({ success: false });
    const mockShowMessage = vi.fn();

    render(
      <MemoryRouter>
        <MessageContext.Provider value={{ showMessage: mockShowMessage }}>
          <AuthContext.Provider value={authValue}>
            <ProfileForm />
          </AuthContext.Provider>
        </MessageContext.Provider>
      </MemoryRouter>
    );

    fireEvent.click(screen.getByRole('button', { name: /Save Changes/i }));

    // Check error handling
    await waitFor(() => {
      expect(mockShowMessage).toHaveBeenCalledWith(['firstName can not be empty', 'lastName can not be empty', 'email can not be empty'], 'warning');
    });
  });
});
