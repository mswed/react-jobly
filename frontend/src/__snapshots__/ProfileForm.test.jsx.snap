// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ProfileForm > matches snapshot 1`] = `
<DocumentFragment>
  <form
    class=""
  >
    <div
      class="mb-3"
    >
      <label
        class="form-label"
        for="username"
      >
        Username
      </label>
      <input
        class="form-control"
        disabled=""
        id="username"
        name="username"
        placeholder="Username"
        value="testuser"
      />
    </div>
    <div
      class="mb-3"
    >
      <label
        class="form-label"
        for="first-name"
      >
        First Name
      </label>
      <input
        class="form-control"
        id="first-name"
        name="firstName"
        placeholder="First Name"
        value=""
      />
    </div>
    <div
      class="mb-3"
    >
      <label
        class="form-label"
        for="last-name"
      >
        Last Name
      </label>
      <input
        class="form-control"
        id="last-name"
        name="lastName"
        placeholder="Last Name"
        value=""
      />
    </div>
    <div
      class="mb-3"
    >
      <label
        class="form-label"
        for="email"
      >
        Email
      </label>
      <input
        class="form-control"
        id="email"
        name="email"
        placeholder="Email"
        value=""
      />
    </div>
    <button
      class="w-100 btn btn-primary"
      type="submit"
    >
      Save Changes
    </button>
  </form>
</DocumentFragment>
`;
